api/spec/requests/spree/api/classifications_controller_spec.rb

```diff
@@ -30,14 +30,24 @@ module Spree
 
       let(:last_product) { taxon.products.last }
 
-      it "can change the order a product" do
+      it "can change the order of a product" do
         classification = taxon.classifications.find_by(product_id: last_product.id)
         expect(classification.position).to eq(3)
         put spree.api_classifications_path, params: { taxon_id: taxon.id, product_id: last_product.id, position: 0 }
         expect(response.status).to eq(200)
         expect(classification.reload.position).to eq(1)
       end
 
+      it "can change the order of a product regardless gaps in positions due discarted products" do
+        taxon.classifications.reload
+        taxon.classifications.second.update_attribute(:position, 3)
+        taxon.classifications.third.update_attribute(:position, 5)
+        classification = taxon.classifications.find_by(product_id: last_product.id)
+        put spree.api_classifications_path, params: { taxon_id: taxon.id, product_id: last_product.id, position: 1 }
+        expect(response.status).to eq(200)
+        expect(classification.reload.position).to eq(3)
+      end
```

Comment:
I think something should be fixed here. I change this code (which I think it's legit) to make a test and it raises an exception:

```
        taxon.classifications.second.update_attribute(:position, 3)
        taxon.classifications.third.update_attribute(:position, 4)
        classification = taxon.classifications.find_by(product_id: last_product.id)
        put spree.api_classifications_path, params: { taxon_id: taxon.id, product_id: last_product.id, position: 3 }
        expect(response.status).to eq(200)
```

Can you please give it a try as well? 

I think that with your solution there's the possiblity that the array could be smaller than the position passed via params.

Reply:
Uhm, the js plugin to reorder the products works from 0 to size - 1, the product used to test has 3 taxons, so, a valid position should be from 0 to 2, isn't it?

Reply:
Ah ok, so it's not possible for the plugin to make a request with position > "number of non-discarded elements", which would raise the exception I had, because `params[:position].to_i` is not an index of that array. 

Makes sense, thanks for the clarification!

Reply:
Maybe we should still make an extra check to avoid this because the request could come from any other place, even not the JS plugin?

Reply:
yes, I agree on that, what do you suggest to return? like a unprocessable_entity with the error in json ?

Reply:
yes, I think that would make sense, thanks!

Reply:
Done! @kennyadsl